FHEM/00_SIGNALduino.pm

```diff
@@ -1477,8 +1477,67 @@ sub SIGNALduino_Write {
 	    $sum	= 12;
   	}
     $msg = $hash->{protocolObject}->PreparingSend_FS20_FHT($id, $sum, $msg);
-  } elsif($fn eq 'k') {   # KOPP_FC
-		$hash->{logMethod}->($name, 4, "$name: Write, cmd $fn sending KOPP_FC");
+  } elsif($fn eq 'k') {   # KOPP_FC   (no outsourcing  in SD_Protocols.pm due to loop and set hash values)
+    $hash->{logMethod}->($name, 4, "$name: Write, cmd $fn sending KOPP_FC");
+    $fn='raw';
+
+    # https://wiki.fhem.de/wiki/Kopp_Allgemein | https://github.com/heliflieger/a-culfw/blob/master/culfw/clib/kopp-fc.c
+    #kr07C2AD1A30CC0F0328
+    #||  ||||  ||    ++-------- Transmitter Code 2
+    #||  ||||  ++-------------- Keycode
+    #||  ++++------------------ Transmitter Code 1
+    #++------------------------ kr wird von der culfw bei Empfang einer Kopp Botschaft als Kennung gesendet
+
+    # $message = "s"
+    #  . $keycodehex
+    #  . $hash->{TRANSMITTERCODE1}
+    #  . $hash->{TRANSMITTERCODE2}
+    #  . $hash->{TIMEOUT}
+    #  . "N";                       # N for do not print messages (FHEM will write error messages to log files if CCD/CUL sends status info
+
+    my $Keycode = substr($msg,1,2);
+    my $TransCode1 = substr($msg,3,4);
+    my $TransCode2 = substr($msg,7,2);
+    my $blkck = 0xAA;
+    my $d;
+
+    ### The device to be sent stores something in own hash. Search for names to access them ###
+    #### The variant with the loop does not require any adjustment in the original Kopp module. ####
+    my $KOPPname;
+    KOPPname:
+    foreach my $d (keys %defs) {
```

Comment:
Da hast Du glaube ich viel Code dupliziert.
Zum einen sollten Internals auch über InternalVal abgerufen werden, zum anderen bin ich jetzt gerade überrascht, dass du alle Definitionen durchsuchst.  Ich empfehle devspec2Array zu verwenden.

Reply:
Wieso ich alles durchsuche, liegt auf der Hand. Ich kenne den Devicenamen nicht. Dieser wird beim senden bzw. Sprung in die SIGNALduino_Write nicht übergeben. Somit muss ich schauen welches Device die Codes hat um mit den Namen zu nehmen. Erst wenn ich den Namen habe, kann ich den Wert auch in dessen Internalval ändern.

InternalVal benötigt auch den Name des Gerätes.
devspec2Array ist neu für mich und auf den Ersten blick ist das auch nur die halbe Miete weil ich dann ebenso auch die Devices noch zusätzlich durchsuchen muss.
--> Die Schleife durchläuft da nur einmal und ich erhalte alles.

Reply:
Devspec2array ist schlecht dokumentiert, da gebe ich dir Recht.
Es hat halt viel mehr Fehlerhandlung enthalten. Z.B. werden Definitionen die ignore enthalten ignoriert.

Alle Bedingungen kann man im Filter Angeben, welche Du hier prüfst.
Am Ende bekommt man allerdings eine Liste von defs zurück.
Aber das müsste man eh zu verlegen, was passiert wenn die Abfrage auf zwei passt.

Reply:
>Aber das müsste man eh zu verlegen, was passiert wenn die Abfrage auf zwei passt.

Wurde simuliert.
Dann hat der User 2 Devices mit der selben Definition angelegt aber unter anderem Namen.

Die Parse verwendet das DEF.  Beide Devices verhalten sich dann identisch. Schaltet Device 1, dann zieht Device 2 gleich und umgekehrt. Daraus schlussfolgernd, benötigt man nur eines von beiden.

Reply:
Nun hattest du mich aber angefixt ;-)
Ich habe die Loop herausgenommen und durch die interne Fuktion Devspec2array ersetzt.
Zuächtlich habe ich den Fall abgefangen wenn jemand mit Absicht oder aus unwissendheit 2 gleiche Devices anlegt, das man dies im Log ausgegeben bekommt.

Reply:
Ja durch das devspec2array wir das ganze deutlich schlanker.
Wenn da nichts gefunden wird, gibts aber auf jeden Fall eine Warnung und bei 0 gefunden devices kommt diese gleiche Logmeldung wie auch für zwei gefundene Resultate Aber ich bin noch am Rätseln wieso wir nicht sendMsg aufrufen und dort das Senden erledigen.

Ich habe jetzt einfach mal angenommen, dass hier gesendet werden soll. Was da Write übergeben wird, weiss ich leider nicht :(

War schon kurz davor, für die write funktion einen test zu schreiben, aber das lohnt sich erst, wenn ich weiss wieso es nicht zu sendMsg gehört :)

Reply:
>Wenn da nichts gefunden wird, gibts aber auf jeden Fall eine Warnung und bei 0 gefunden devices kommt diese gleiche Logmeldung wie auch für zwei gefundene Resultate Aber ich bin noch am Rätseln wieso wir nicht sendMsg aufrufen und dort das Senden erledigen.

Ja das ist erstmal richtig.
Der Dreh & Angelpunkt ist vermutlich schon weiter "vorn" in FHEM. Die IOWrite Funktion übergibt an das Modul auch, sowohl wenn der Empfänger "offline" ist bzw. disconnectet. Das bräuchte nicht sein. Da könnte man schon abbrechen.
Es müsste als letzter Schritt in der IOWrite eine Prüfung rein, ob der Empfänger wirklcih aktiv ist und dann handeln oder Abbrechen um im Modul bei der Write Funktion anzukommen. (Wenn ich das richtig nachvollzogen habe)

Mein Test belief sich drarauf, das SIGNALduino Modul auf verbos 5 zu stellen und bei einem Device einen Befehl abzusetzen obwohl der Empfänger offline ist. Da kommen wir durch bis zu Write Funktion, obwohl wir ja nicht senden könnten ;-)

Reply:
IOWrite kann ja von allem möglichen aufgerufen werden.

FHEM läuft nur, weil viele Annahmen meist zutreffen. Robuster Code sieht ein wenig anders aus. Aber gut, wir machen das hobby mäßig und dafür läuft der Kram auch gut.
Wer rumbastelt, der weiss ja auch selbst, dass er was gebastelt hat :)

Dem Prüfen, über IOWrite, ob der Empfänger aktiv ist, da fehlt mir gerade das Hintergrundwissen, woher man wissen will ob ein Empfänger aktiv ist.
Die Frage die noch nicht für mich beantwortet ist, ist ja eher ob da nicht Code in sendMSG gehört.

Was für Befehle werden denn mittels IOWrite da versuchsweise an die Hardware gegeben? 
Wie ich diesen Legacy Kram doch liebe :)

Reply:
> Die Frage die noch nicht für mich beantwortet ist, ist ja eher ob da nicht Code in sendMSG gehört.

Wenn du von der Anpassung beim senden redest, denke ich nicht. Die Position erscheint mir als richtig.

> Dem Prüfen, über IOWrite, ob der Empfänger aktiv ist, da fehlt mir gerade das Hintergrundwissen, woher man wissen will ob ein Empfänger aktiv ist.

Da müsste geprüft werden, das das Device an der Stelle nicht disconnected / closed ist. Schlussfolgerung wäre, es ist Sendebereit.

Reply:
Ich gehe nochmal auf deine Aussage drauf ein

>Die Frage die noch nicht für mich beantwortet ist, ist ja eher ob da nicht Code in sendMSG gehört.

Bis zu sendMSG kommen wir nicht. Beispiel senden Kopp. Das modul übergibt an die FHEM IOWrite, welche dann an das SIGNALduino Modul in die Write Funktion übergibt. Hier passen wir den Sendebefehl an und übergeben dann an "Set raw". Zu "Set sendmsg" kommen wir nicht. Du musst dort keine Anpassungen o. Ähnliches vornehmen.
... nicht das es zu Missverständnissen oder zu einer falschen Denkweise kommt.

Reply:
Gefühlt war ich an diesem Punkt schon einmal, dass irgend ein Legacy Modul IOWrite aufgerufen hat und etwas direkt an die Hardware übergeben wollte. 
Tolle Idee sage ich dazu nur, aber so ist das Kopp Modul wohl leider entwickelt.

Jedenfalls, erinnere ich mich, dass dann das Senden auch über sendMSG funktionieren sollte, was ja durchaus logisch klingt, dass der Sendebefehl auch ein Protokoll senden kann und damit der Code mitverwendet werden kann, hat IOWrite die Daten an den sendMSG Teil übergeben.
Davon ist vermutlich nur deshalb nicht mehr so viel zu sehen, weil in das logische Modul dann eingebaut wurde den set Befehl aufzurufen und nicht iowrite, denn schreiben direkt auf die Hardware ist ja Käse.

Set raw für einen Sendebfehl zu verwenden ist auch noch ungünstig, weil dort das sendworking Flag nicht gesetzt wird. Das ist aber nicht ganz unwichtig, um weitere Befehle abwarten zu können. Im Endeffekt, tricksen wir das komplette Modul aus. Ich glaube sogar die maximale Sendezeitberechnung bekommt davon dann nichts mehr mit.

Wie sieht denn am Ende das "raw" Kommando aus? Ich Tippe mal auf SR; .......

Reply:
> Wie sieht denn am Ende das "raw" Kommando aus? Ich Tippe mal auf SR; .......

SN;R=13;N=4;D=07C2AD0000CC0F0302000000000000;

Was soll deine ausführliche Erklärung soeben sagen? Nein geht nicht so umzusetzen? Ich verstehe nur einen Teil davon.

Reply:
Ich glaube es ist besser, das Senden über den sendMSG zu realisieren. So wie auch hier:
https://github.com/RFD-FHEM/RFFHEM/blob/745d3e6377b7854f61cfa20e5877fc224d418359/FHEM/00_SIGNALduino.pm#L1465-L1480

Dann passt es auch zum Codepfad und den Logmeldungen, denn der Rückgabewert durchläuft auf jeden Fall
SIGNALduino_CheckSendRawResponse. 
Sendworking wird zum Glück für den SR, SN etc Befehl gesetzt, da macht es auch keinen Unterschied.
Wenn dann aber ggf. zum Senden noch Register umgestellt werden müssen ist es in sendMsg besser aufgehoben, dann die Register passend zum Protokoll zu setzen. Meinst Du nicht auch?

Reply:
>Wenn dann aber ggf. zum Senden noch Register umgestellt werden müssen ist es in sendMsg besser aufgehoben, 

Wir wählen vorher den rfmode aus. Somit wird das Register gesetzt und ich muss nicht beim senden das Register setzen.

>Sendworking wird zum Glück für den SR, SN etc Befehl gesetzt, da macht es auch keinen Unterschied.

Also kann es beim senden via RAW bleiben?

Reply:
Wer wählt den rfmode denn vor dem Senden aus?

Wenn ich jetzt beispielsweise ein Kopp Gerät habe das ich nicht ständig empfangen muss, aber befehle dorthin senden möchte, dann feuere ich den Sendebefehl ab und irgendie wird das attribut rfmode geändert?

Ich glaube bei Somfy war das, da wird über das Protokoll die Frequenz zum Senden angepasst.
Wäre das hier nicht auch ratsam?


Bezüglich senden, geht FS20  und FHT den Weg über sendMsg.
Mir wäre daran gelegen es einheitlich zu haben. Den Vorteil es nicht darüber laufen zu lassen habe ich bislang nicht erkannt.

Reply:
>Wer wählt den rfmode denn vor dem Senden aus?

Die User, welche nur ein Protokoll fest verarbeiten. Ich denke, du lässt dich hier täuschen, das du gern ASK weiterhin empfangen möchtest und anderes senden wie Bsp Kopp. Das Attribut ist wie beim CUL der Modus was man haben möchte. Wenn du nun mit "gemischten" Empfang beginnst, das geht nicht gut mit einem cc1101. Ich kann auch nicht 2 unterschiedliche Sensoren von unterschiedlicher Modulation empfangen. somit muss ich vorher mich entscheiden via rfmode.

>Ich glaube bei Somfy war das, da wird über das Protokoll die Frequenz zum Senden angepasst.
Wäre das hier nicht auch ratsam?

Bei dieser Umsetzung sehe ich dort keinen Vorteil weil ich stets fest einen Modus vorher einstellen.

>Den Vorteil es nicht darüber laufen zu lassen habe ich bislang nicht erkannt.

Du kannst sofort und einfach die Nachrichten ohne zu wandeln wieder heraussenden. Wenn wir nun anfangen das auf sendMsg umzustellen so wird der Code mehr.